aocd/models.py

```diff
@@ -262,10 +256,14 @@ def _repr_pretty_(self, p, cycle):
             p.text(template.format(type(self).__name__, self, hex(id(self))))
 
     def _coerce_val(self, val):
-        if isinstance(val, float) and val.is_integer():
+        if isinstance(val, np_complex) and val.imag == 0:
+            val = val.real
+        if isinstance(val, np_float):
+            val = float(val)
+        if isinstance(val, np_timedelta) or (isinstance(val, (float)) and val.is_integer()):
             log.warning("coerced value %r for %d/%02d", val, self.year, self.day)
             val = int(val)
-        if isinstance(val, int):
+        if isinstance(val, (int, np_int)):
```

Comment:
Shouldn't this if-statement be removed? (Not related to this review, it's the same issue on master.) We should always make sure to convert `val` to `str` in this function. The easiest fix would be to just change the return statement to `return str(val)`.

Reply:
I had the opposite instinct, actually - we're using this function to format our submission in order to prevent submitting bad answers (ie, the "4.0" != "4" problem that was solved by the original coercion code). My thought was potentially creating an allowlist of types we'll accept for `str` casting, because not all types will cast sanely:

```python3
Python 3.11.0rc1 (main, Aug 12 2022, 10:02:14) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
>>> x = numpy.timedelta64(123)
>>> str(x)
'123 generic time units'
>>> isinstance(x, numpy.integer)
True
```

It might be worth explicitly defending against submitting unknown types as answers. I have no strong opinion either way, though.

Reply:
Ah, I see. Maybe casting non-`str` types first to `int` and then `str` would work slightly better:
```python3
if not isinstance(val, str):
    if isinstance(val, float) and val.is_integer():
        log.warning("coerced value %r for %d/%02d", val, self.year, self.day)
    val = int(val)
    val = str(val)
return val
```